server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java

```diff
@@ -39,15 +40,23 @@ public abstract class SearchProgressListener {
      * @param skippedShards The list of skipped shards.
      * @param clusters The statistics for remote clusters included in the search.
      * @param fetchPhase <code>true</code> if the search needs a fetch phase, <code>false</code> otherwise.
+     * @param timeProvider absolute and relative time provider for this search
      **/
-    protected void onListShards(List<SearchShard> shards, List<SearchShard> skippedShards, Clusters clusters, boolean fetchPhase) {}
+    protected void onListShards(
+        List<SearchShard> shards,
+        List<SearchShard> skippedShards,
+        Clusters clusters,
+        boolean fetchPhase,
+        TransportSearchAction.SearchTimeProvider timeProvider
+    ) {}
 
     /**
      * Executed when a shard returns a query result.
      *
-     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards} )}.
+     * @param shardIndex  The index of the shard in the list provided by {@link SearchProgressListener#onListShards} )}.
+     * @param queryResult
      */
-    protected void onQueryResult(int shardIndex) {}
+    protected void onQueryResult(int shardIndex, QuerySearchResult queryResult) {}
```

Comment:
I was able to get all the logic for cluster status updates by using the SearchProgressListener. Only three changes were needed:

1) Adding QuerySearchResult to this method - in order to detect if a search has timedOut on a shard (cluster)
2) Adding  TransportSearchAction.SearchTimeProvider timeProvider to the onListShards method - in order to be able to set `took` values on a per cluster basis.
3) Adding the method override in `SearchDfsQueryThenFetchAsyncAction` of:

```java
    @Override
    protected void onShardGroupFailure(int shardIndex, SearchShardTarget shardTarget, Exception exc) {
        progressListener.notifyQueryFailure(shardIndex, shardTarget, exc);
    }
```

`SearchQueryThenFetchAsyncAction.java` already does this, so it was likely an oversight that the Dfs class didn't have it.

Reply:
Update: I had to add Cluster update logic to the minimize_roundtrips=false CCSActionListener as well to handle the edge case where the SearchShards API call to do remote can-match and shard listing returns no shards at all. See https://github.com/elastic/elasticsearch/pull/98457/files/0ac26078563544e9811d59c1e64e072a13fce5fd#diff-24921eca27bf57c7b3d8d3ba7f02acbf6449d91e43e49bbd9df6377843adb19aR677